### ISSUE TITLE
Add rule S6270: Policies authorizing public access to resources are security-sensitive

### ISSUE THREAD

**Issue by pierre-loup-tristant-sonarsource** [SonarSource/SonarJS#3449]
# What

The goal of this rule is to detect Python code using the AWS CDK that creates resource-based policies granting public access.

This has been previously implemented for Python. See [SONARPY-1144](https://sonarsource.atlassian.net/browse/SONARPY-1144).
# What

# Detection logic

## `PolicyStatement`

Policies are defined by classes such as `Policy` or `ManagedPolicy`. These classes accept a `PolicyDocument`, which contains multiple `PolicyStatement`s. This `PolicyStatement` class should be the main focus of this rule.

### Constructor

The [constructor](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.PolicyStatement.html) receives a single PolicyStatementProps parameters. The following checks should be applied to the parameters of the PolicyStatementProps object that the constructor receives

Status | `principals` | `effect`
-- | -- | --
Noncompliant |  <ul><li> `aws_cdk.aws_iam.StarPrincipal`</li><li>`aws_cdk.aws_iam.AnyPrincipal` </li><li>`aws_cdk.aws_iam.ArnPrincipal` with `"*"` as first parameter (arn) | <ul><li>`iam.Effect.ALLOW`</li><li>Not set
Compliant | Anything other than <ul><li> `aws_cdk.aws_iam.StarPrincipal`</li><li>`aws_cdk.aws_iam.AnyPrincipal` </li><li>`aws_cdk.aws_iam.ArnPrincipal` with `"*"` as first parameter (arn) | Any

**Example**
```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement({
    sid: "AllowAnyPrincipal",
    effect: iam.Effect.ALLOW, // secondary location
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AnyPrincipal()] // Noncompliant
})

new iam.PolicyStatement({
    sid: "DenyAnyPrincipal",
    effect: iam.Effect.DENY,
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AnyPrincipal()] // Compliant
})

new iam.PolicyStatement({
    sid: "AllowAccountRootPrincipal",
    effect: iam.Effect.ALLOW,
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AccountRootPrincipal()] // Compliant
})
```

### `fromJson`

Apart from the constructor, the static method `PolicyStatement.fromJson()` may also be used to create a `PolicyStatement`. It takes an object as first parameter, which should be checked as follows (using JSONPath syntax):

Status | `$.Principal` | `$.Effect`
-- | -- | --
Noncompliant | <ul><li>`*` </li><li> `"AWS": ["*"]` (single element)  </li><li>`"AWS": ["999999999999","*"]` (contains `*` amount other elements) | <ul><li>`"Allow"` (case insensitive) </li><li>`iam.Effect.ALLOW`
Compliant | Anything other than <ul><li>`*` </li><li> `"AWS": ["*"]` (single element)  </li><li>`"AWS": ["999999999999","*"]` (contains `*` amount other elements) | Any

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement.fromJson({
    "Sid": "AnyPrincipal",
    "Effect": "Allow", // secondary location
    "Action": ["kms:*"], 
    "Resource": "*",
    "Principal": {  "AWS": [ "*" ] } // Noncompliant
})

new iam.PolicyStatement.fromJson({
    "Sid": "AllowAccountRootPrincipal",
    "Effect": "Allow",
    "Action": ["kms:*"], 
    "Resource": "*",
    "Principal": { "AWS" : new iam.AccountRootPrincipal().arn } // Compliant
})
```

### `fromJson`

`PolicyDocument` also has a static method `PolicyDocument.fromJson()` which will also indirectly create `PolicyStatements`. This method also takes a object, where `$.Statement` contains a `PolicyStatement[]` which is equal to the input of `PolicyStatement.fromJson()`.

The checks applied to `PolicyStatement.fromJson()` must be applied to each element of `$.Statement`.

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

const policyDocument = {
    "Version": "2012-10-17",
    "Statement": [{
        "Sid": "AllowAnyPrincipal",
        "Effect": "Allow", // secondary location
        "Action": ["kms:*"], 
        "Resource": "*",
        "Principal": {  "AWS": [ "*", ], // Noncompliant
    },
    {
        "Sid": "AccountRootPrincipal",
        "Effect": "Allow",
        "Action": ["kms:*"], 
        "Resource": "*",
        "Principal": { "AWS" : new iam.AccountRootPrincipal().arn } // Compliant
    }]        
}

const policyDocumentFromJson = iam.PolicyDocument.fromJson(policyDocument)
```

## RSPEC

This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1311) contains the message and highlighting guidelines.

**Comment by loris-s-sonarsource** [comment]
✅ Spec LGTM!
